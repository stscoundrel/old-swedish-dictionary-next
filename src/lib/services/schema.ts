import { capitalize } from 'lib/utils/strings'
import { slugifyLetter } from 'lib/utils/slugs'
import { getWordLink, getLetterLink } from 'lib/utils/links'
import { DictionaryEntry } from 'lib/models/dictionary'
import { Breadcrumb } from 'lib/utils/breadcrumbs'

interface SchemaListItem {
  '@type': string,
  position: number
  name: string,
  item: string,
}

interface SchemaDefinition {
  '@context': string,
  '@type': string,
  '@id'?: string,
  name?: string,
  description?: string,
  itemListElement?: SchemaListItem[],
  inDefinedTermSet?: string,
}

const getDefinedTermSetData = (content: DictionaryEntry[]): SchemaDefinition => {
  const letter = {
    letter: content[0].headword.charAt(0),
    slug: slugifyLetter(content[0].headword.charAt(0)),
  }

  return {
    '@context': 'https://schema.org/',
    '@type': 'DefinedTermSet',
    '@id': getLetterLink(letter),
    name: `Old Swedish Dictionary - Letter ${letter.letter.toUpperCase()}`,
    description: `Old Swedish words starting with letter ${letter.letter.toUpperCase()}`,
  }
}

const getDefinedTermData = (content: DictionaryEntry): SchemaDefinition => ({
  '@context': 'https://schema.org/',
  '@type': 'DefinedTerm',
  '@id': getWordLink(content),
  name: `Old Swedish Dictionary - ${capitalize(content.headword)}`,
  description: content.definitions[0],
  inDefinedTermSet: process.env.NEXT_PUBLIC_SITE_URL,
})

const getBreadcrumbListData = (content: Breadcrumb[]): SchemaDefinition => {
  const siteUrl = process.env.NEXT_PUBLIC_SITE_URL

  const listItems: SchemaListItem[] = content.map(({ label, url }, index) => ({
    '@type': 'ListItem',
    position: index + 1,
    name: label,
    item: siteUrl + url,
  }))

  return {
    '@context': 'https://schema.org/',
    '@type': 'BreadcrumbList',
    itemListElement: listItems,
  }
}

const getDefault = (): SchemaDefinition => {
  const siteUrl = process.env.NEXT_PUBLIC_SITE_URL

  return {
    '@context': 'https://schema.org/',
    '@type': 'DefinedTermSet',
    '@id': `${siteUrl}`,
    name: 'Old Swedish Dictionary',
    description: 'Old Swedish words with Swedish definitions',
  }
}

export const getWordSchema = (entry: DictionaryEntry): string => {
  const schema = getDefinedTermData(entry)
  return JSON.stringify(schema)
}

export const getLetterSchema = (entries: DictionaryEntry[]): string => {
  const schema = getDefinedTermSetData(entries)
  return JSON.stringify(schema)
}

export const getBreadcrumbsSchema = (breadcrumbs: Breadcrumb[]): string => {
  const schema = getBreadcrumbListData(breadcrumbs)
  return JSON.stringify(schema)
}

export const getDefaultSchema = (): string => {
  const schema = getDefault()
  return JSON.stringify(schema)
}
