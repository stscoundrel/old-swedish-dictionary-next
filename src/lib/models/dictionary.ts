import { DictionaryEntry as RawDictionaryEntry } from 'old-swedish-dictionary'

export type OriginalDictionaryEntry = RawDictionaryEntry

export interface DictionaryEntry extends RawDictionaryEntry {
    slug: string,
}
