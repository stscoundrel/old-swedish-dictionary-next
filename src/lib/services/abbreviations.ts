import OSAbbreviations from 'old-swedish-dictionary-abbreviations'
import { abbreviate } from 'abbreviatrix'
import { DictionaryEntry } from 'old-swedish-dictionary'

export interface Abbreviation{
    abbreviation: string,
    explanation: string
}

export const getAbbreviations = ({ definitions }: DictionaryEntry): Abbreviation[] => {
  const combinedAbbreviations: Abbreviation[] = []
  const abbreviationSet = new Set()

  definitions.forEach((definition) => {
    const abbreviations = OSAbbreviations.findWorksAndAuthors(definition)
    abbreviations.forEach((explanation, abbreviation) => {
      if (!abbreviationSet.has(abbreviation)) {
        abbreviationSet.add(abbreviation)
        combinedAbbreviations.push({ abbreviation, explanation })
      }
    })
  })

  return combinedAbbreviations;
}

/**
 * Add abbr tags to content with explanations.
 */
export const addAbbreviationsToContent = (
  content: string,
  abbreviations: Abbreviation[],
): string => {
  let result = content

  abbreviations.forEach(({ abbreviation, explanation }) => {
    result = abbreviate(abbreviation, explanation, result)
  })

  return result
}

export const getAllAbbreviations = (): Abbreviation[] => {
  const abbrs = OSAbbreviations.getWorksAndAuthorsMapping()
  return Array
    .from(abbrs, ([abbreviation, explanation]) => ({ explanation, abbreviation }))
    .sort((a, b) => a.abbreviation.localeCompare(b.abbreviation))
}
