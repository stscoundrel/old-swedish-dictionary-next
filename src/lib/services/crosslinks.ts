import { getCrosslinks as getAllCrossLinks, Crosslink, getOldSwedishCrosslinks } from 'scandinavian-dictionary-crosslinker'
import { DictionaryEntry } from 'lib/models/dictionary'

export type { Crosslink } from 'scandinavian-dictionary-crosslinker'

let crossLinkCache: Record<string, Crosslink[]> | null = null

const populateCrossLinks = () => {
  if (!crossLinkCache) {
    crossLinkCache = getAllCrossLinks()
  }

  return crossLinkCache
}

export const getCrossLinks = (entry: DictionaryEntry): Crosslink[] => {
  const crosslinks = populateCrossLinks()
  if (Object.prototype.hasOwnProperty.call(crosslinks, entry.slug)) {
    return getOldSwedishCrosslinks(entry.slug)
  }

  return []
}
