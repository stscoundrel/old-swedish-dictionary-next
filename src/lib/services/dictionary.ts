import { getDictionary } from 'old-swedish-dictionary'
import { slugifyWord, slugifyLetter } from '../utils/slugs'
import { OriginalDictionaryEntry, DictionaryEntry } from '../models/dictionary'

let cachedDictionary: DictionaryEntry[] | null = null

export interface AlphabetLetter {
  letter: string,
  slug: string,
}

const addSlugs = (words: OriginalDictionaryEntry[]): DictionaryEntry[] => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.headword).toLowerCase()

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = (): DictionaryEntry[] => {
  if (cachedDictionary) return cachedDictionary

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  cachedDictionary = formattedWords

  return formattedWords
}

export const getByLetter = (letter: string): DictionaryEntry[] => {
  const words = getAllWords()
  const byLetter = words.filter((entry) => (
    entry.headword.charAt(0).toLowerCase() === letter.toLowerCase()))

  return byLetter
}

export const getWord = (word: string): DictionaryEntry => (
  getAllWords().filter((entry) => entry.slug === word)[0]
)

export const getAlphabet = (): AlphabetLetter[] => {
  // Outputted from dictionary src with scripts/output-alphabet.js
  const letters = [
    'a',
    'b',
    'c',
    'd',
    'e',
    'f',
    'g',
    'h',
    'i',
    'j',
    'k',
    'l',
    'm',
    'n',
    'o',
    'p',
    'q',
    'r',
    's',
    't',
    'u',
    'v',
    'y',
    'ä',
    'ö',
    'þ',
  ]

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}
