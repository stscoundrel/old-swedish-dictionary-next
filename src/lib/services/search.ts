import { markWords } from 'markari'
import { DictionaryEntry } from 'lib/models/dictionary'

export interface SearchResult {
  headword: string,
  slug: string,
  foundIn: string[],
}

export type Criteria = 'headword' | 'definitions'

const formatResults = (
  results: DictionaryEntry[],
  search: string,
  criteria: Criteria[],
): SearchResult[] => {
  const formattedResults = results.map((result) => {
    const foundIn: string[] = []

    if (criteria.includes('definitions')) {
      result.definitions.forEach((definition) => {
        if (definition.toLowerCase().includes(search.toLowerCase())) {
          foundIn.push(markWords(search, definition))
        }
      })
    }

    if (foundIn.length === 0) {
      const highlightedHeadword = markWords(search, result.headword)
      foundIn.push(`In headword: ${highlightedHeadword}`)
    }

    return {
      headword: result.headword,
      slug: result.slug,
      foundIn,
    }
  })

  return formattedResults
}

export const searchDictionary = (
  search: string,
  dictionary: DictionaryEntry[],
  criteria: Criteria[] = ['headword', 'definitions'],
): SearchResult[] => {
  const filteredSearch = search.toLowerCase()

  const results = dictionary.filter((entry) => {
    let matchesSearch = false

    if (criteria.includes('headword')) {
      if (entry.headword.toLowerCase().includes(filteredSearch)) {
        matchesSearch = true
      }

      if (entry.slug.includes(filteredSearch)) {
        matchesSearch = true
      }
    }

    if (criteria.includes('definitions')) {
      entry.definitions.forEach((definition) => {
        if (definition.toLowerCase().includes(filteredSearch)) {
          matchesSearch = true
        }
      })
    }

    return matchesSearch
  })

  const formattedResult = formatResults(results, search, criteria)

  return formattedResult
}

export default searchDictionary
