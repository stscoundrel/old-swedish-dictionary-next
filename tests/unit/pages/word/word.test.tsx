import ReactDOM from 'react-dom/client'
import Word, { getStaticProps, getStaticPaths } from 'pages/word/[word]'
import renderer from 'react-test-renderer'
import { getAlphabet } from 'lib/services/dictionary'
import { Crosslink, DictionarySource } from 'scandinavian-dictionary-crosslinker'

const mockHandler = jest.fn()

/**
 * Mock router
 */
jest.mock('next/router', () => ({
  useRouter() {
    return {
      locale: undefined,
      defaultLocale: undefined,
      asPath: '/test',
      back: mockHandler,
    }
  },
}))

const entry = {
  headword: 'afköra',
  partOfSpeech: ['vb'],
  grammaticalAspect: 'v.',
  information: '',
  definitions: [
    ', vräka. hafwi wald han afköra  GS 43 (1416?). af ty ahren (för areno) han landboen afkörer ib.',
  ],
  alternativeForms: ['-bortdrifva'],
  slug: 'afkora',
}

const letter = {
  letter: 'a',
  slug: 'a',
}

const abbreviations = [
  {
    abbreviation: 'GS',
    explanation: 'Några Gambla Stadgar [såsom bihang till Biärköa-Rätten utg. af J. Hadorph]. 1687.',
  },
]

const crosslinks: Crosslink[] = [
  {
    url: 'https://cleasby-vigfusson-dictionary.vercel.app/word/fadir',
    source: DictionarySource.OldNorse,
  },
  {
    url: 'https://old-icelandic.vercel.app/word/fadir',
    source: DictionarySource.OldIcelandic,
  },
  {
    url: 'https://old-norwegian-dictionary.vercel.app/word/fadir',
    source: DictionarySource.OldNorwegian,
  },
]

describe('Word page: render & usage', () => {
  test('Does not crash', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(
      <Word
        entry={entry}
        similarEntries={[]}
        letters={getAlphabet()}
        letter={letter}
        abbreviations={abbreviations}
        crosslinks={crosslinks}
      />,
    )
  })

  test('Matches snapshot', () => {
    const tree = renderer.create(
      <Word
        entry={entry}
        similarEntries={[]}
        letters={getAlphabet()}
        letter={letter}
        abbreviations={abbreviations}
        crosslinks={crosslinks}
      />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Returns null if entry is unavailable', () => {
    const tree = renderer.create(
      <Word
        entry={null}
        similarEntries={[]}
        letters={getAlphabet()}
        letter={letter}
        abbreviations={abbreviations}
        crosslinks={crosslinks}
      />,
    ).toJSON()
    expect(tree).toBeNull()
  })

  test('Back button works', async () => {
    const tree = renderer.create(
      <Word
      entry={entry}
      similarEntries={[]}
      letters={getAlphabet()}
      letter={letter}
      abbreviations={abbreviations}
      crosslinks={crosslinks}
    />,
    )

    // Click back btn.
    await renderer.act(async () => {
      expect(mockHandler).not.toHaveBeenCalled()
      await tree.root.findByProps({ text: 'Back' }).props.action()

      // Assert mockrouter received a push.
      expect(mockHandler).toHaveBeenCalled()
      expect(mockHandler.mock.calls.length).toBe(1);
    })
  })
})

describe('Word page: data fetching', () => {
  test('getStaticPaths works', async () => {
    const result = await getStaticPaths()

    expect(result.fallback).toBe('blocking')

    // Should've build initial pages.
    expect(result.paths.length).toBe(5964)
  })

  test('getStaticProps works', async () => {
    const expected = {
      props: {
        entry,
        similarEntries: [],
        letters: getAlphabet(),
        letter,
        abbreviations,
        crosslinks: [],
      },
    }

    const result = await getStaticProps({ params: { word: 'afkora' } })

    expect(result).toEqual(expected)
  })

  test('getStaticProps finds similar entries based on slug', async () => {
    const expected = [
      {
        headword: 'fosterfadhir',
        partOfSpeech: ['nn'],
        grammaticalAspect: '',
        information: '',
        definitions: [
          'L: fosterfader, försörjare. jhesus war thera fosteradher til thera lifs bärningh  MP 4: 29.',
        ],
        alternativeForms: ['fooster- GU C 20 s. 367) ,'],
        slug: 'fosterfadhir-2',
      },
    ]

    const result = await getStaticProps({ params: { word: 'fosterfadhir' } })
    expect(result.props.similarEntries).toEqual(expected)
  })

  test('getStaticProps returns 404 redirect for unkown words', async () => {
    const expected = {
      props: {},
      notFound: true,
    }

    const result = await getStaticProps({ params: { word: 'loremipsum' } })

    expect(result).toEqual(expected)
  })
})
