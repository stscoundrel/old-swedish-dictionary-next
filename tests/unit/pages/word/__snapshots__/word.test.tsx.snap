// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Word page: render & usage Matches snapshot 1`] = `
[
  <header>
    <div
      className="topBarSpacer"
    />
    <nav
      className="section"
    >
      <div
        className="topbar container"
      >
        <a
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          <img
            alt="To home"
            data-nimg="1"
            decoding="async"
            height={30}
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            src="/_next/image?url=%2Ffavicon-48x48.png&w=64&q=75"
            srcSet="/_next/image?url=%2Ffavicon-48x48.png&w=32&q=75 1x, /_next/image?url=%2Ffavicon-48x48.png&w=64&q=75 2x"
            style={
              {
                "color": "transparent",
              }
            }
            width={30}
          />
        </a>
        <div
          aria-label="Open menu"
          className="hamburger"
          onClick={[Function]}
          role="button"
        >
          <span
            className="part"
          />
          <span
            className="part"
          />
          <span
            className="part"
          />
        </div>
      </div>
      <div
        className="content null container"
      >
        <ul
          className="list"
        >
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/a"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              a
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/b"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              b
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/c"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              c
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/d"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              d
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/e"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              e
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/f"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              f
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/g"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              g
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/h"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              h
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/i"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              i
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/j"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              j
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/k"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              k
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/l"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              l
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/m"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              m
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/n"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              n
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/o"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              o
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/p"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              p
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/q"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              q
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/r"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              r
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/s"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              s
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/t"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              t
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/u"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              u
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/v"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              v
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/y"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              y
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/ae"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              ä
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/oe"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              ö
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/th"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              þ
            </a>
          </li>
        </ul>
        <form
          className="form"
          onSubmit={[Function]}
        >
          <input
            aria-label="Search"
            className="input"
            name="search"
            onChange={[Function]}
            type="search"
            value=""
          />
          <button
            className="button button--secondary button--no-margs"
            name="submit"
            type="submit"
          >
            Search
          </button>
        </form>
      </div>
    </nav>
    <nav
      className="section"
    >
      <div
        className="container"
      >
        <a
          className="link"
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          Old Swedish Dictionary
        </a>
        <a
          className="link"
          href="/letter/a"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          Letter A
        </a>
        <a
          className="link"
          href="/word/afkora"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          Afköra
        </a>
        <script
          dangerouslySetInnerHTML={
            {
              "__html": "{"@context":"https://schema.org/","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Old Swedish Dictionary","item":"undefined/"},{"@type":"ListItem","position":2,"name":"Letter A","item":"undefined/letter/a"},{"@type":"ListItem","position":3,"name":"Afköra","item":"undefined/word/afkora"}]}",
            }
          }
          type="application/ld+json"
        />
      </div>
    </nav>
  </header>,
  <main
    className="container"
  >
    <article
      className="section"
    >
      <header>
        <h1
          lang="non-swe"
        >
          Afköra
        </h1>
        <small
          className="subHeading"
        >
          Old Swedish Dictionary - 
          afköra
        </small>
        <p>
          Meaning of Old Swedish word 
          <em>
            "
            afköra
            "
          </em>
           in Swedish.
        </p>
        <p>
          As defined by K.F Söderwall's dictionary of Medieval Swedish:
        </p>
      </header>
      <dl
        className="definitionList"
      >
        <dt>
          <strong>
            afköra
          </strong>
        </dt>
        <dd
          lang="swe"
        >
           , vräka. hafwi wald han afköra  GS 43 (1416?). af ty ahren (för areno) han landboen afkörer ib.
        </dd>
      </dl>
      <p>
        <strong>
          Part of speech:
        </strong>
         
        <em>
          vb
        </em>
      </p>
      <p>
        <strong>
          Grammatical aspect:
        </strong>
         
        <em>
          v.
        </em>
      </p>
      <strong>
        Alternative forms or notes:
      </strong>
      <ul
        className="list"
      >
        <li>
          -bortdrifva
        </li>
      </ul>
      <p>
        Possible runic inscription in 
        <em>
          Medieval Futhork
        </em>
        :
        <span
          className="rune"
        >
          ᛆᚠᚴᚯᚱᛆ
        </span>
        <br />
        <small>
          Medieval Runes were used in Sweden from 12th to 17th centuries.
        </small>
      </p>
    </article>
    <a
      className="button"
      onClick={[Function]}
    >
      Back
    </a>
    <div
      aria-label="Back to top"
      className="backToTop"
      onClick={[Function]}
      role="button"
    >
      ↑
    </div>
  </main>,
  <footer
    className="section"
  >
    <div
      className="container"
    >
      <section
        className="section"
      >
        <h2>
          About
        </h2>
        <p>
          Based on "Ordbok Öfver svenska medeltids-språket" by K.F Söderwall.
        </p>
        <p>
          It is the largest Old Swedish dictionary. The original volumes were published in 1884 - 1918, with additional supplement added in 1953 - 1973.
        </p>
      </section>
      <section
        className="section"
      >
        <h4>
          Old Swedish language
        </h4>
        <p>
          Old Swedish developed from Old East Norse, the eastern dialect of Old Norse, at the end of the Viking Age.
        </p>
        <p>
          Early Old Swedish was spoken from about 1225 until about 1375, and Late Old Swedish was spoken from about 1375 until about 1526.
        </p>
        <p>
          Also known as "Medieval Swedish", or "Fornsvenska" in modern Swedish
        </p>
      </section>
      <section
        className="section"
      >
        <h4>
          Old Norse language
        </h4>
        <p>
          Old Norse was a North Germanic language that was spoken by inhabitants of Scandinavia and their overseas settlements from about the 7th to the 15th centuries.
        </p>
        <p>
          Also known as "the viking language", "Old Nordic", or "Old Scandinavian"
        </p>
      </section>
      <div
        className="navs"
      >
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Dictionary project
          </h4>
          <ul>
            <li>
              <a
                href="https://github.com/stscoundrel/old-swedish-dictionary-next"
                rel="noopener noreferrer"
                target="_blank"
              >
                Source code
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/old-swedish-dictionary"
                rel="noopener noreferrer"
                target="_blank"
              >
                Data source
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/old-swedish-dictionary-builder"
                rel="noopener noreferrer"
                target="_blank"
              >
                Data source builder
              </a>
            </li>
          </ul>
        </nav>
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Related dictionary projects
          </h4>
          <ul>
            <li>
              <a
                href="https://old-norwegian-dictionary.vercel.app/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Dictionary of the Old Norwegian Language
              </a>
            </li>
            <li>
              <a
                href="https://cleasby-vigfusson-dictionary.vercel.app/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Cleasby and Vigfusson Old Norse Dictionary
              </a>
            </li>
            <li>
              <a
                href="https://old-icelandic.vercel.app/"
                rel="noopener noreferrer"
                target="_blank"
              >
                A Concise Dictionary of Old Icelandic
              </a>
            </li>
          </ul>
        </nav>
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Quick links
          </h4>
          <ul
            className="navColumns"
          >
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/a"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                a
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/b"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                b
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/c"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                c
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/d"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                d
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/e"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                e
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/f"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                f
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/g"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                g
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/h"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                h
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/i"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                i
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/j"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                j
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/k"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                k
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/l"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                l
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/m"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                m
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/n"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                n
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/o"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                o
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/p"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                p
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/q"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                q
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/r"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                r
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/s"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                s
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/t"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                t
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/u"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                u
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/v"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                v
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/y"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                y
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/ae"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                ä
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/oe"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                ö
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/th"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                þ
              </a>
            </li>
          </ul>
        </nav>
      </div>
      <small
        className="copyright"
      >
        Copyright © 2022 - 2021
        <br />
        <a
          href="https://github.com/stscoundrel"
          rel="noopener noreferrer"
          target="_blank"
        >
          Sampo Silvennoinen / StScoundrel
        </a>
      </small>
    </div>
  </footer>,
]
`;
